File: apps/chat-ios/AmazonChimeSDKMessagingDemo/messaging-websocket/utils/DefaultSigV4.swift
//
//  DefaultSigV4.swift
//  AmazonChimeSDKMessagingDemo
//
//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
//  SPDX-License-Identifier: Apache-2.0
//

import AWSCore

let X_AMZ_ALGORITHM: String = "X-Amz-Algorithm"
let X_AMZ_CREDENTIAL: String = "X-Amz-Credential"
let X_AMZ_DATE: String = "X-Amz-Date"
let X_AMZ_EXPIRES: String = "X-Amz-Expires"
let X_AMZ_SIGNED_HEADERS: String = "X-Amz-SignedHeaders"
let X_AMZ_SECURITY_TOKEN: String = "X-Amz-Security-Token"

public class DefaultSigV4: SigV4 {
    private var credentials: ChimeUserCredentials
    private var region: String

    init(credentials: ChimeUserCredentials, region: String) {
        self.credentials = credentials
        self.region = region
    }

    public func signUrl(method: String,
                 scheme: String,
                 serviceName: String,
                 hostname: String,
                 path: String,
                 payload: String,
                 queryParams: [String: String]) -> String {

        let now = getDateTimeString()
        let today = getDateString(dateTimeString: now)

        let algorithm = "AWS4-HMAC-SHA256"
        let signedHeaders = "host"
        let canonicalHeaders = "host:\(hostname.lowercased())\n"
        let credentialScope = "\(today)/\(region)/\(serviceName)/aws4_request"

        var params: [String : String] = [
            X_AMZ_ALGORITHM: algorithm,
            X_AMZ_CREDENTIAL: encodeURI(content: "\(credentials.accessKeyId)/\(credentialScope)"),
            X_AMZ_DATE: now,
            X_AMZ_EXPIRES: "10",
            X_AMZ_SIGNED_HEADERS: signedHeaders,
            X_AMZ_SECURITY_TOKEN: encodeURI(content: credentials.sessionToken)
        ]

        if !queryParams.isEmpty {
            for (key, value) in queryParams {
                params[key] = encodeURI(content: value)
            }
        }

        let sorted = params.sorted(by: { $0.0 < $1.0 })
        var canonicalQueryString = ""
        for (key, value) in sorted {
            if !canonicalQueryString.isEmpty {
                canonicalQueryString.append("&")
            }
            canonicalQueryString.append(key)
            canonicalQueryString.append("=")
            canonicalQueryString.append(value)
        }

        let canonicalRequest = method + "\n" +
                               path + "\n" +
                               canonicalQueryString + "\n" +
                               canonicalHeaders + "\n" +
                               signedHeaders + "\n" +
                               sha256(content: payload)

        let hashedCanonicalRequest: String = sha256(content: canonicalRequest)
        let stringToSign = algorithm + "\n" +
                           now + "\n" +
                           credentialScope + "\n" +
                           hashedCanonicalRequest
        print("DefaultSigV4 signURL stringToSign: \(stringToSign)")

        let signingKey = getSignatureKey(key: credentials.secretAccessKey, date: today, regionName: region, serviceName: serviceName)
        let signature = hmac(data: stringToSign, key: signingKey).hexEncodedString()

        let finalParams = "\(canonicalQueryString)&X-Amz-Signature=\(signature)"
        return "\(scheme)://\(hostname)\(path)?\(finalParams)"
    }

    private func getDateTimeString() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        dateFormatter.timeZone = (NSTimeZone(name: "UTC")! as TimeZone)
        dateFormatter.locale = Locale(identifier: Locale.current.languageCode!)
        return dateFormatter.string(from: Date())
    }

    private func getDateString(dateTimeString: String) -> String {
        return String(dateTimeString[..<dateTimeString.firstIndex(of: "T")!])
    }

    private func encodeURI(content: String) -> String {
        let allowedCharacterSet = CharacterSet(charactersIn: "!*'();:@&=+$,/?%#[] ").inverted
        return content.addingPercentEncoding(withAllowedCharacters: allowedCharacterSet)!
    }

    private func sha256(content: String) -> String {
        return AWSSignatureSignerUtility.hexEncode(AWSSignatureSignerUtility.hashString(content))
    }

    private func getSignatureKey(key: String, date: String, regionName: String, serviceName: String) -> Data {
        let kSecret = Data("AWS4\(key)".utf8)
        let kDate = hmac(data: date, key: kSecret)
        let kRegion = hmac(data: regionName, key: kDate)
        let kService = hmac(data: serviceName, key: kRegion)
        let kSigning = hmac(data: "aws4_request", key: kService)
        return kSigning
    }

    func hmac(data: String, key: Data) -> Data {
        return AWSSignatureSignerUtility.sha256HMac(with: Data(data.utf8), withKey: key)
    }
}

File: apps/chat-android/messaging-websocket/src/main/java/com/amazonaws/services/chime/sdk/messaging/utils/DefaultSigV4.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.amazonaws.services.chime.sdk.messaging.utils

import com.amazonaws.services.chime.sdk.messaging.session.ChimeUserCredentials
import com.amazonaws.services.chime.sdk.messaging.utils.logger.Logger
import java.lang.StringBuilder
import java.net.URLEncoder
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class DefaultSigV4(
    private val logger: Logger,
    private val credentials: ChimeUserCredentials,
    private val region: String
) : SigV4 {
    private val TAG = "DefaultSigV4"

    override fun signURL(
        method: String,
        scheme: String,
        serviceName: String,
        hostname: String,
        path: String,
        payload: String,
        queryParams: Map<String, List<String>>?
    ): String {
        val now = getDateTimeString()
        val today = getDateString(now)

        val algorithm = "AWS4-HMAC-SHA256"
        val signedHeaders = "host"
        val canonicalHeaders = "host:${hostname.toLowerCase(Locale.getDefault())}\n"
        val credentialScope = "$today/$region/$serviceName/aws4_request"

        var params = mutableMapOf<String, MutableList<String>>().apply {
            set("X-Amz-Algorithm", mutableListOf(algorithm))
            set(
                "X-Amz-Credential", mutableListOf(
                    encodeURIComponent("${credentials.accessKeyId}/$credentialScope")
                )
            )
            set("X-Amz-Date", mutableListOf(now))
            set("X-Amz-Expires", mutableListOf("10"))
            set("X-Amz-SignedHeaders", mutableListOf("host"))
            credentials.sessionToken?.let {
                set("X-Amz-Security-Token", mutableListOf(encodeURIComponent(it)))
            }
        }

        if (queryParams != null) {
            for ((key, values) in queryParams) {
                val encodedKey = encodeURIComponent(key)
                values.sorted()
                values.forEach {
                    if (!params.containsKey(encodedKey)) {
                        params[encodedKey] = mutableListOf()
                    }
                    params[encodedKey]?.add(encodeURIComponent(it))
                }
            }
        }

        val canonicalQueryStringBuilder = StringBuilder()
        // Query string has to be sorted by parameters
        params = params.toSortedMap()
        params.forEach { (key, values) ->
            values.forEach { value ->
                if (canonicalQueryStringBuilder.isNotEmpty()) {
                    canonicalQueryStringBuilder.append("&")
                }
                canonicalQueryStringBuilder.append(key)
                canonicalQueryStringBuilder.append("=")
                canonicalQueryStringBuilder.append(value)
            }
        }
        val canonicalQueryString = canonicalQueryStringBuilder.toString()

        val canonicalRequest =
            method +
                    "\n" +
                    path +
                    "\n" +
                    canonicalQueryString +
                    "\n" +
                    canonicalHeaders +
                    "\n" +
                    signedHeaders +
                    "\n" +
                    hex(sha256(payload))

        val hashedCanonicalRequest = hex(sha256(canonicalRequest))

        val stringToSign =
            "AWS4-HMAC-SHA256\n" +
                    now +
                    "\n" +
                    today +
                    "/" +
                    region +
                    "/" +
                    serviceName +
                    "/aws4_request\n" +
                    hashedCanonicalRequest

        val signingKey = getSignatureKey(credentials.secretAccessKey, today, region, serviceName)

        val signature = hex(hmac(stringToSign, signingKey))

        val finalParams = "$canonicalQueryString&X-Amz-Signature=$signature"

        return "$scheme://$hostname$path?$finalParams"
    }

    private fun getDateTimeString(): String {
        val format = SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'", Locale.getDefault())
        format.timeZone = TimeZone.getTimeZone("UTC")
        return format.format(Date())
    }

    private fun getDateString(dateTimeString: String): String {
        return dateTimeString.substring(0, dateTimeString.indexOf('T'))
    }

    private fun encodeURIComponent(uriComponent: String): String {
        return URLEncoder.encode(uriComponent, "utf-8")
    }

    private fun sha256(string: String): ByteArray {
        return MessageDigest
            .getInstance("SHA-256")
            .digest(string.toByteArray())
    }

    private fun hmac(data: String, key: ByteArray): ByteArray {
        val algorithm = "HmacSHA256"
        val mac = Mac.getInstance(algorithm)
        mac.init(SecretKeySpec(key, algorithm))
        return mac.doFinal(data.toByteArray())
    }

    private fun hex(bytes: ByteArray): String {
        return bytes.fold("", { str, it -> str + "%02x".format(it) })
    }

    private fun getSignatureKey(key: String, date: String, regionName: String, serviceName: String): ByteArray {
        val kSecret = "AWS4$key".toByteArray()
        val kDate = hmac(date, kSecret)
        val kRegion = hmac(regionName, kDate)
        val kService = hmac(serviceName, kRegion)
        val kSigning = hmac("aws4_request", kService)
        return kSigning
    }
}
